src/fetch.js

```diff
@@ -262,75 +256,76 @@ const getSyncData = async (
         ? response.sync_token
         : aggregatedResponse.sync_token;
     }
+
+    // Handle pagination
     if (response.pagination_token) {
-      try {
-        return await getSyncData(
-          url,
-          config,
-          { pagination_token: response.pagination_token },
-          responseKey,
-          aggregatedResponse,
-          0 // Reset retries for each call
-        );
-      } catch (error) {
-        if (retries < config.httpRetries) {
-          const timeToWait = 2 ** retries * 100;
-          //Retry attempt ${retries + 1} after pagination token error. Waiting for ${timeToWait} ms...
-          await waitFor(timeToWait);
-          return await getSyncData(
-            url,
-            config,
-            { pagination_token: response.pagination_token },
-            responseKey,
-            aggregatedResponse,
-            retries + 1
-          );
-        } else {
-          throw new Error(`Failed to fetch sync data after ${config.httpRetries} retry attempts due to invalid pagination token.`);
-        }
-      }
+      return handlePagination(url, config, response.pagination_token, responseKey, aggregatedResponse);
     }
 
+    // Handle sync tokens for final sync call
     if (response.sync_token) {
-      /**
-       * To make final sync call and concatenate the result if found any during on fetch request.
-       */
-      const aggregatedSyncToken = syncToken.filter(item => item !== undefined);
-      let SyncRetryCount = 0;
-      for (const token of aggregatedSyncToken) {
-        let syncResponse;
-        try {
-          syncResponse = await fetchCsData(
-            url,
-            config,
-            (query = { sync_token: token })
+      aggregatedResponse = await processSyncTokens(url, config, aggregatedResponse, syncToken);
+    }
+
+    return aggregatedResponse;
+  } catch (error) {
+    throw new Error(`Failed to fetch sync data: ${error.message}`);
+  }
+};
+
+const processSyncTokens = async (url, config, aggregatedResponse, syncToken) => {
+  // Remove Duplicate/ undefined sync tokens [if any] before iterating.
+  const aggregatedSyncToken = [...new Set(syncToken.filter(item => item !== undefined))];
+  // Clear the syncToken array so further processing won’t include already handled tokens.
+  syncToken.length = 0;
+
+  let lastError = null;
+
+  for (const token of aggregatedSyncToken) {
+    let syncResponse;
+    let SyncRetryCount = 0;
+     // Attempt to fetch data using the token, with retries.
+     while (SyncRetryCount <= config.httpRetries) {
+      try {
+        syncResponse = await fetchCsData(url, config, { sync_token: token });
+        break; 
+      } catch (error) {
+        SyncRetryCount++;
+        lastError = error;
+        if (SyncRetryCount <= config.httpRetries) {
+          const delay = 2 ** (SyncRetryCount - 1) * 100;
+          console.info(`Hold on... Retrying...`);
```

Comment:
could you please check the reporter API for logging instead of console.info or warn

https://www.gatsbyjs.com/docs/tutorial/creating-a-source-plugin/part-4/#reporter-api